### ISSUE TITLE
Version information unavailable to nmap while libbitcoin node is busy

### ISSUE THREAD

**Issue by JasonCoombs** [libbitcoin/libbitcoin-node#393]
This:
https://github.com/libbitcoin/libbitcoin-node/issues/392#issuecomment-436084464

When the node is busy communicating with other nodes (inbound or outbound) this error occurs:

```
Nmap scan report for 127.0.0.1
Host is up, received conn-refused (0.00028s latency).
Scanned at 2018-11-05 14:13:32 HST for 10s

PORT     STATE SERVICE REASON
8333/tcp open  bitcoin syn-ack
| bitcoin-info:
|_  ERROR: Failed to extract version information
Final times for host: srtt: 278 rttvar: 3805  to: 100000
```

When nmap is the first and only inbound connection, version information is extracted and displayed:

```
Host is up, received conn-refused (0.00015s latency).
Scanned at 2018-11-05 14:12:39 HST for 0s

PORT     STATE SERVICE REASON
8333/tcp open  bitcoin syn-ack
| bitcoin-info:
|   Timestamp: 2018-11-06T00:12:39
|   Network: main
|   Version: 0.7.0
|   Node Id: e8fad242281281e2
|   Lastblock: 42561
|_  User Agent: /libbitcoin:4.0.0/
Final times for host: srtt: 149 rttvar: 3772  to: 100000
```

**Comment by JasonCoombs** [comment]
but you don't mind if I fix it, right?

**Comment by evoskuil** [comment]
This is a non-bug. See linked comments.

**Comment by evoskuil** [comment]
Yes, I would mind, as it would slow down the node.

**Comment by JasonCoombs** [comment]
we are intentionally disabling user agent remote discovery while there is network traffic? could you point out the area of the source code relevant to this? thanks.

**Comment by evoskuil** [comment]
No, we aren't. We are dropping the slowest peers in a batch connect once the first peer in the batch completes the handshake.

**Comment by evoskuil** [comment]
You can verify this is the behavior by setting the batch size to 1. Then it will only drop after timeout or error.

**Comment by JasonCoombs** [comment]
but nmap is an inbound peer, and I've configured support for up to 100 inbound connections. why should my nmap client be dropped?

**Comment by JasonCoombs** [comment]
thanks, I'll keep looking at it... the batch size of 1 did not change the behavior.

here's what it looked like when nmap was invoked repeatedly prior to the first outbound connection, after which point nmap was unable to receive and display the user agent detail:

```
00:24:14.282406 INFO [node] Node is started.
00:24:18.689605 INFO [network] Connected inbound channel [127.0.0.1:51955] (1)
00:24:22.980997 INFO [network] Connected inbound channel [127.0.0.1:51960] (1)
00:24:59.965202 INFO [network] Connected inbound channel [127.0.0.1:51974] (1)
00:25:03.158453 INFO [network] Connected inbound channel [127.0.0.1:51980] (1)
00:25:05.272159 INFO [network] Connected inbound channel [127.0.0.1:51984] (1)
00:25:24.387614 INFO [network] Connected inbound channel [127.0.0.1:51989] (1)
00:25:25.738666 INFO [network] Connected inbound channel [127.0.0.1:51993] (1)
00:25:26.859299 INFO [network] Connected inbound channel [127.0.0.1:51997] (1)
00:25:27.912764 INFO [network] Connected inbound channel [127.0.0.1:52001] (1)
00:25:31.793318 INFO [network] Connected outbound channel [31.208.8.175:8333] (1)
00:25:33.464169 INFO [node] 0x70000a532000 Header #66000 [00000000071d7e8a0f4895e60c1073df9311d65a85244be1ee6369c9506281af] from [31.208.8.175:8333] (139*, 1).
00:25:33.611107 INFO [blockchain] 0x70000a532000 Organized blocks [62594-62594]
```

**Comment by JasonCoombs** [comment]
is there a reason we can't leave this issue open until there's a resolution to it?

**Comment by evoskuil** [comment]
Sure, closing it was based on the incorrect belief it was related to batch timeouts.

Also, your console output above is not the debug log. Check the debug log file in verbose configuration.

**Comment by evoskuil** [comment]
As I corrected, it's not related to batching if it's inbound.

**Comment by JasonCoombs** [comment]
I increased the nmap NSE script timeout to 60 seconds.

This timeout can be set here (default path on Mac OS):

`/usr/local/share/nmap/scripts/bitcoin-info.nse`

line 50, changed to 60000 ms timeout:

`local bcoin = bitcoin.Helper:new(host, port, { timeout = 60000 })`

frequently it takes just over 15 seconds for libbitcoin-node to respond, and I've seen this happen several times when there are no outbound connections established yet.

```
NSE: Finished bitcoin-info against 127.0.0.1:8333.
Completed NSE at 16:09, 15.13s elapsed
Nmap scan report for 127.0.0.1
Host is up, received conn-refused (0.00035s latency).
Scanned at 2018-11-05 16:09:41 HST for 15s
```

**Comment by JasonCoombs** [comment]
why would a libbitcoin-node instance which has not yet started doing anything other than wait for batch outbound connections to be established with peers require 15 seconds to respond to a new inbound peer connection?

**Comment by JasonCoombs** [comment]
I've seen timers being set to wake up the proactor message queue processing but I don't yet understand why those timers are necessary. is this an example of the performance problem caused by relying on those timers, and should there have been a different solution to whatever problem causes queue processing to halt for which the timers may have been a workaround?

**Comment by JasonCoombs** [comment]
They can be. But I think there's a bug in our implementation and use of async connect which causes all network threadpool result handlers to wait or queue and not get processed until the `connect_timeout_seconds` have elapsed. I'll modify my setting to something larger (but below my 60 second nmap timeout) and try nmap again to see whether libbitcoin ignores my pending inbound connection while it is waiting for a timeout on the outbound connection. That shouldn't be happening if we are using async socket notifications correctly, particularly for connect.

Also, see https://github.com/libbitcoin/libbitcoin-node/issues/396
